metrics/src/recorder.rs

```diff
@@ -1,13 +1,102 @@
-const INITIALIZED: usize = 2;
+    /// The global Recorder instance with a `once_cell`-like API
+    pub struct RecorderOnceCell {
+        recorder: UnsafeCell<Option<&'static dyn Recorder>>,
+        state: AtomicUsize,
+    }
+
+    /// The recorder is uninit and can be set
+    const UNINITIALIZED: usize = 0;
+    /// The recorder is currently being initialized
+    const INITIALIZING: usize = 1;
+    /// The recorder has been initialized successfully and can be read
+    const INITIALIZED: usize = 2;
+
+    impl RecorderOnceCell {
+        pub const fn new() -> Self {
+            Self { recorder: UnsafeCell::new(None), state: AtomicUsize::new(UNINITIALIZED) }
+        }
+
+        #[cfg(atomic_cas)]
+        pub fn set(&self, recorder: &'static dyn Recorder) -> Result<(), SetRecorderError> {
+            // Acquire the lock because the write below must not be reordered above the CAS
+            match self.state.compare_exchange(
+                UNINITIALIZED,
+                INITIALIZING,
+                Ordering::Acquire,
+                Ordering::Relaxed,
+            ) {
+                Ok(UNINITIALIZED) => {
+                    unsafe {
+                        // SAFETY: Access is unique because we CASed the state to INITIALIZING above
+                        self.recorder.get().write(Some(recorder));
+                    }
+                    // Release the lock, others can now read it - but not write
+                    self.state.store(INITIALIZED, Ordering::Release);
+                    Ok(())
+                }
+                _ => Err(SetRecorderError(())),
+            }
+        }
+
+        /// # Safety
+        /// The caller must guarantee that no reader has read the state before we do this and then
+        /// reads the recorder after another writer has written to it after us.
+        pub unsafe fn clear(&self) {
+            // Set the state to `UNINIT` to allow the next writer to write again.
+            // This is not a problem for readers since their `&'static` refs will remain
+            // valid forever.
+            // Use SeqCst to be safer, but Relaxed should work just as well
+            self.state.store(UNINITIALIZED, Ordering::SeqCst);
+        }
+
+        pub fn try_load(&self) -> Option<&'static dyn Recorder> {
+            if self.state.load(Ordering::Acquire) != INITIALIZED {
+                None
+            } else {
+                // SAFETY: Thanks to `Acquire` above we make sure that this doesn't get
+                // reordered above this and therefore no writer is here
+                unsafe { self.recorder.get().read() }
+            }
+        }
+
+        pub unsafe fn set_racy(
+            &self,
+            recorder: &'static dyn Recorder,
+        ) -> Result<(), SetRecorderError> {
+            match self.state.load(Ordering::Acquire) {
+                UNINITIALIZED => {
+                    // SAFETY: Access is unique because we CASed the state to INITIALIZING above
```

Comment:
This safety comment doesn't sound right, access is unique because the safety precondition of `set_racy` is that this isn't ran concurrently.